sharkfuser/include/fusilli/graph.h

```diff
@@ -158,6 +150,71 @@ class Graph : public INode {
   }
 
 private:
+  // CleanupCacheDirectory removes a sub-directory from the main cache directory
+  // if it's empty. When used as a base class, C++ destructor ordering
+  // (explained below) ensures that the directory cleanup in
+  // CleanupCacheDirectory destructor will happen after any CacheFiles member
+  // variables have been destroyed. If member destructors (~M1, ~M2) are called
+  // inside ~B; the compiler will still destroy members afterward, leading to
+  // double-destruction (UB).
+  //
+  // Destructor ordering example:
+  //   struct A   { ~A()  {std::cout << "A";} };
+  //   struct M1  { ~M1() {std::cout << "M1, ";} };
+  //   struct M2  { ~M2() {std::cout << "M2, ";} };
+  //
+  //   struct B : A {
+  //       M1 m1;
+  //       M2 m2;
+  //       ~B() { std::cout << "B, "; }
+  //   };
+  //   // output -> "B, M2, M1, A"
+  struct CleanupCacheDirectory {
+    std::filesystem::path cacheDir;
+    explicit CleanupCacheDirectory(std::filesystem::path dir)
+        : cacheDir(std::move(dir)) {}
+
+    ~CleanupCacheDirectory() {
+      // This likely indicates the instance in question has been moved from.
+      if (cacheDir.empty()) {
+        return;
+      }
+
+      if (std::filesystem::exists(cacheDir) &&
+          std::filesystem::is_empty(cacheDir)) {
+        std::filesystem::remove(cacheDir);
+      }
+    }
+  };
+
+  // Holds cached assets generated from a `generateCompiledArtifact` call. If
+  // `CacheFiles` are set to be removed RAII based removal will be tied to this
+  // object.
+  struct CachedAssets : CleanupCacheDirectory {
```

Comment:
Unless this is tied to the `Graph` class, please move to `extras.h` where rest of `CacheFile` also live.